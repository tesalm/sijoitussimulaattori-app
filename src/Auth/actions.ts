import { to } from 'await-to-js';
import firebase from 'react-native-firebase';
import { Dispatch } from 'redux';

import { t } from '../assets/i18n';
import { getCurrentUser, getUserData, signInAnonymously, signOut, upsertUserData } from '../firestore';
import { User, UserAuth } from '../models';
import { createDefaultUser } from '../util/users';
import { UserData } from './../models/user.model';

export enum ActionType {
  RestoreLoginSuccess = '[Login] Restore Login Success',
  RestoreLoginImpossible = '[Login] Restore Login Impossible',
  LoginRequest = '[Login] Login Request',
  LoginSuccess = '[Login] Login Success',
  LoginFailure = '[Login] Login Failure',
  LogoutRequest = '[Login] Logout Request',
  FetchUserDataSuccess = '[Login] Fetch User Data Success',
  FetchUserDataFailure = '[Login] Fetch User Data Failure',
  DeleteCurrentUserRequest = '[Login] Delete Current User Request',
  DeleteCurrentUserSuccess = '[Login] Delete Current User Success',
  DeleteCurrentUserFailure = '[Login] Delete Current User Failure',
}

export type AuthAction =
  | RestoreLoginSuccess
  | RestoreLoginImpossible
  | LoginRequest
  | LoginSuccess
  | LoginFailure
  | LogoutRequest
  | FetchUserDataSuccess
  | FetchUserDataFailure
  | DeleteCurrentUserRequest
  | DeleteCurrentUserSuccess
  | DeleteCurrentUserFailure;

export class LoginRequest {
  readonly type = ActionType.LoginRequest;
  constructor() {
    return { type: this.type };
  }
}

export class LoginSuccess {
  readonly type = ActionType.LoginSuccess;
  constructor(public user: User) {
    return { type: this.type, user };
  }
}

export class LoginFailure {
  readonly type = ActionType.LoginFailure;
  constructor(public error: Error) {
    return { type: this.type, error };
  }
}

export class LogoutRequest {
  readonly type = ActionType.LogoutRequest;
  constructor() {
    return { type: this.type };
  }
}

export class FetchUserDataSuccess {
  readonly type = ActionType.FetchUserDataSuccess;
  constructor(public user: User) {
    return { type: this.type, user };
  }
}

export class FetchUserDataFailure {
  readonly type = ActionType.FetchUserDataFailure;
  constructor(public userDataError: Error) {
    return { type: this.type, userDataError };
  }
}

export class RestoreLoginSuccess {
  readonly type = ActionType.RestoreLoginSuccess;
  constructor(public userAuth: UserAuth) {
    return { type: this.type, userAuth };
  }
}

export class RestoreLoginImpossible {
  readonly type = ActionType.RestoreLoginImpossible;
  constructor() {
    return { type: this.type };
  }
}

export class DeleteCurrentUserRequest {
  readonly type = ActionType.DeleteCurrentUserRequest;
  constructor() {
    return { type: this.type };
  }
}

export class DeleteCurrentUserSuccess {
  readonly type = ActionType.DeleteCurrentUserSuccess;
  constructor() {
    return { type: this.type };
  }
}

export class DeleteCurrentUserFailure {
  readonly type = ActionType.DeleteCurrentUserFailure;
  constructor(public error: Error) {
    return { type: this.type, error: this.error };
  }
}
const restorePreviousLogin = () => async (dispatch: Dispatch<AuthAction>) => {
  // Wait the completion of the Firebase initialization and catch the logged in user
  // when after it.
  const unsubscripe = firebase
    .auth()
    .onAuthStateChanged((userAuth: UserAuth) => {
      if (userAuth) {
        // Previous user can be restored.
        dispatch(new RestoreLoginSuccess(userAuth));
        fetchUserData(dispatch, userAuth);
      } else {
        // There is no previous user to restore.
        dispatch(new RestoreLoginImpossible());
      }

      // Catch only the first authentication state change event.
      unsubscripe();
    });
};

// This function should be used only for "old" users.
const fetchUserData = async (
  dispatch: Dispatch<AuthAction>,
  userAuth: UserAuth
) => {
  let userData: UserData = {};

  try {
    userData = (await getUserData(userAuth.uid)) || userData;
  } catch (err) {
    dispatch(new FetchUserDataFailure(err));
    return; // <- Exit on error.
  }

  const user: User = {
    ...userAuth,
    ...userData,
  };

  dispatch(new FetchUserDataSuccess(user));
};

const login = () => async (dispatch: Dispatch<AuthAction>) => {
  dispatch(new LoginRequest());

  const [err, fsUser] = await to(signInAnonymously());

  if (err || !fsUser) {
    return dispatch(new LoginFailure(err || Error('User not found.')));
  }

  const userAuth: UserAuth = {
    uid: fsUser.user.uid,
  };

  // It should be now that the login was successfull.
  // The rest of this function will handle user data.
  dispatch(new LoginSuccess(userAuth));

  let userData: UserData = {};

  if (fsUser.additionalUserInfo && fsUser.additionalUserInfo.isNewUser) {
    // Create new user.
    userData = createDefaultUser();

    try {
      await upsertUserData(userAuth.uid, userData);
    } catch (err) {
      console.error('Error creating defaut user:' + err); // TODO: Error handling needed? Maybe dispatch failure notification action
    }
  } else {
    // Fetch data for "old" user.
    fetchUserData(dispatch, userAuth);
    return; // <- !
  }

  const user: User = {
    ...userAuth,
    ...userData,
  };

  // Assign user data for new user. The user might got nothing as UserData, if something
  // went wrong on the user creation initialization.
  dispatch(new FetchUserDataSuccess(user));
};

const logout = () => async (dispatch: Dispatch<LogoutRequest>) => {
  const [err] = await to(signOut());

  if (err) {
    console.error('Error logging out');
  } else {
    dispatch(new LogoutRequest());
  }
};

const deleteCurrentUser = () => async (dispatch: Dispatch<AuthAction>) => {
  dispatch(new DeleteCurrentUserRequest());

  const currentUser = getCurrentUser();

  if (!currentUser) {
    dispatch(
      new DeleteCurrentUserFailure(new Error(t('Auth.NoAuthenticatedUser')))
    );
    return;
  }

  const [err] = await to(currentUser.delete());

  if (err) {
    dispatch(new DeleteCurrentUserFailure(err));
  } else {
    dispatch(new DeleteCurrentUserSuccess());
  }
};

export { login, logout, deleteCurrentUser, restorePreviousLogin };
