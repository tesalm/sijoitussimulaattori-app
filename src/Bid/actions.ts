import { Dispatch } from 'redux';
import { BidInfo } from './reducer';

export enum ActionType {
  UpdateAction = '[Bid] Update action',
  UpdateBidLevel = '[Bid] Update bid level',
  UpdateSumOfStocks = '[Bid] Update sum of stocks',
  UpdateSelectedPortfolio = '[Bid] Update selected portfolio',
  SetSymbol = '[Bid] Set symbol',
}

export type BidAction =
  | UpdateAction
  | UpdateBidLevel
  | UpdateSumOfStocks
  | UpdateSelectedPortfolio
  | SetSymbol;

export class UpdateAction {
  readonly type = ActionType.UpdateAction;
  constructor(public actionType: string) {
    return { type: this.type, actionType };
  }
}

export class UpdateBidLevel {
  readonly type = ActionType.UpdateBidLevel;
  constructor(public bidLevel: number) {
    return { type: this.type, bidLevel };
  }
}

export class UpdateSumOfStocks {
  readonly type = ActionType.UpdateSumOfStocks;
  constructor(public sumOfStocks: number) {
    return { type: this.type, sumOfStocks };
  }
}

export class UpdateSelectedPortfolio {
  readonly type = ActionType.UpdateSelectedPortfolio;
  constructor(public selectedPortfolio: string) {
    return { type: this.type, selectedPortfolio };
  }
}

export class SetSymbol {
  readonly type = ActionType.SetSymbol;
  constructor(public symbol: string) {
    return { type: this.type, symbol };
  }
}

const updateAction = (action: string) => async (
  dispatch: Dispatch<BidAction>
) => {
  dispatch(new UpdateAction(action));
};

const updateBidLevel = (bidLevel: number) => async (
  dispatch: Dispatch<BidAction>
) => {
  dispatch(new UpdateBidLevel(bidLevel));
};

const updateSumOfStocks = (sumOfStocks: number) => async (
  dispatch: Dispatch<BidAction>
) => {
  dispatch(new UpdateSumOfStocks(sumOfStocks));
};

const updateSelectedPortfolio = (selectedPortfolio: string) => async (
  dispatch: Dispatch<BidAction>
) => {
  dispatch(new UpdateSelectedPortfolio(selectedPortfolio));
};

const saveBidForm = (
  action: string,
  bidLevel: number,
  sumOfStocks: number,
  selectedPortfolio: string,
  symbol: string
) => async (dispatch: Dispatch<BidAction>) => {
  dispatch(new UpdateAction(action));
  dispatch(new UpdateBidLevel(bidLevel));
  dispatch(new UpdateSumOfStocks(sumOfStocks));
  dispatch(new UpdateSelectedPortfolio(selectedPortfolio));
  dispatch(new SetSymbol(symbol));
};

export {
  updateAction,
  updateBidLevel,
  updateSumOfStocks,
  updateSelectedPortfolio,
  saveBidForm,
};
