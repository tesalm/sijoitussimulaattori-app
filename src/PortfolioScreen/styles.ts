import { StyleSheet } from 'react-native';
import { scale, verticalScale } from 'react-native-size-matters';

import { Colors } from '../App/colors';

const portfolioContainerStyles = StyleSheet.create({
  basicInfo: {
    margin: 0,
    height: 130,
  },
  diagram: {
    margin: 0,
    height: 200,
  },
  buttonContainer: {
    margin: 0,
    height: 50,
  },
  holdings: {
    margin: 0,
    flex: 1,
  },
});

const portfolioStyles = StyleSheet.create({
  titleStyle: {
    fontWeight: 'bold',
    fontSize: 17,
    marginBottom: scale(8),
    color: Colors.baseColor,
  },
  valueHeader: {
    fontSize: 12,
    textAlign: 'justify',
    marginBottom: verticalScale(0),
    color: Colors.baseColor,
  },
  valueHeaderMiddle: {
    fontSize: 12,
    textAlign: 'justify',
    marginTop: verticalScale(20),
    marginLeft: scale(16),
    color: Colors.baseColor,
  },

  value: {
    fontSize: 14,
    fontWeight: 'bold',
    textAlign: 'justify',
    marginTop: verticalScale(-2),
    marginBottom: verticalScale(4),
    color: Colors.baseColor,
  },
  valueGreen: {
    fontSize: 14,
    fontWeight: 'bold',
    textAlign: 'justify',
    marginTop: verticalScale(-2),

    marginBottom: verticalScale(4),
    color: Colors.greenPercent,
  },
  valueRed: {
    fontSize: 14,
    fontWeight: 'bold',
    textAlign: 'justify',
    marginTop: verticalScale(-2),

    marginBottom: verticalScale(4),
    color: Colors.redPercent,
  },
  valueMiddle: {
    fontSize: 14,
    fontWeight: 'bold',
    textAlign: 'justify',
    marginTop: verticalScale(-2),
    marginLeft: scale(16),
    marginBottom: verticalScale(8),
    color: Colors.baseColor,
  },
  valueHeaderRightSide: {
    fontSize: 12,
    marginTop: verticalScale(20),
    marginLeft: scale(30),
    color: Colors.baseColor,
  },
  valueHeaderRightSideHoldings: {
    fontSize: 12,
    marginTop: verticalScale(30),
    marginLeft: scale(30),
    color: Colors.baseColor,
  },

  titleText: {
    fontSize: 17,
    fontWeight: 'bold',
    marginLeft: scale(8),

    color: Colors.baseColor,
  },
  holdingsContainer: {
    display: 'flex',
    flexDirection: 'row',
    marginBottom: 8,
  },
  portfolioinfo: {
    display: 'flex',
    flexDirection: 'row',
  },
  portfolioInfoSmallerComp: {
    flexDirection: 'column',
    flex: 1,
  },
  loading: {
    height: '100%',
    justifyContent: 'center',
    alignItems: 'center',
  },
  accordionContainer: {
    borderTopWidth: 1.5,
  },
  holdingsSubTitle: {
    fontSize: 15,
    marginBottom: scale(8),
    marginTop: scale(8),
    color: Colors.baseColor,
    fontWeight: 'bold',
  },
  holdingsLogoView: {
    marginTop: 0,
    flexDirection: 'row',
    marginBottom: scale(12),
  },
  holdingsSubLogoView: {
    marginTop: 0,
    flexDirection: 'row',
    marginBottom: scale(12),
  },
  holdingsSubTitleView: {
    flex: 0.9,
  },
  holdingsArrowView: {
    flex: 0.1,
    marginTop: scale(5),
  },
});

export { portfolioContainerStyles, portfolioStyles };
