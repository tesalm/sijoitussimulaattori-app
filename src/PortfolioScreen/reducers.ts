import cloneDeep from 'lodash/cloneDeep';

import { ActionType, PortfolioAction } from './actions';

export interface PortfolioStock {
  uid: string;
  amount: number;
  avgPrice: number;
}
export interface Portfolio {
  uid: string;
  portfolioId: string;
  balance: number;
  revenue: number;
  totalMarketValue: number;
  ownerId: string;
  stocks: PortfolioStock[];
}

export interface SinglePortfolio {
  portfolio?: Portfolio;
  loading: boolean;
  error?: Error;
  refreshing: boolean;
}

const initialState: SinglePortfolio = {
  portfolio: undefined,
  loading: false,
  error: undefined,
  refreshing: false,
};

export const portfolioReducer = (
  state: SinglePortfolio = initialState,
  action: PortfolioAction
): SinglePortfolio => {
  switch (action.type) {
    case ActionType.RequestPortfolioBegin:
      return {
        ...cloneDeep(state),
        loading: true,
        error: undefined,
        refreshing: true,
      };
    case ActionType.RequestPortfolioSuccess:
      return {
        ...state,
        portfolio: action.portfolio,
        loading: false,
        error: undefined,
        refreshing: false,
      };
    case ActionType.RequestPortfolioFailure:
      return {
        ...cloneDeep(state),
        loading: false,
        error: action.error,
        refreshing: false,
      };

    default:
      return state;
  }
};
