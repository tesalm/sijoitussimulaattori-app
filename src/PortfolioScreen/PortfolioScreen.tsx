import debounce from 'lodash/debounce';
import React from 'react';
import { RefreshControl, ScrollView, Text } from 'react-native';
import { Card } from 'react-native-elements';
import { NavigationScreenProps } from 'react-navigation';
import { connect } from 'react-redux';
import { bindActionCreators, Dispatch } from 'redux';

import { Colors } from '../App/colors';
import { cardButtonStyles, cardStyles } from '../App/styles';
import CardButton from '../general/cardButton';
import { IconNames } from '../general/icon';
import { getIntraday, getStockMetadata, getStocks, refreshIntraday } from '../MarketScreen/actions';
import { Stock } from '../MarketScreen/reducer';
import { RouteName } from '../navigation/routes';
import { cancelTransaction, getPortfolioData, getTransactions } from '../PortfolioList/actions';
import { PortfolioStock, SinglePortfolio } from '../PortfolioList/reducer';
import { RootState } from '../redux/reducers';
import { t } from './../assets/i18n';
import { Holdings } from './components/Holdings';
import { OpenTransactions } from './components/OpenTransactions';
import { PortfolioInfo } from './components/PortfolioInfo';

export interface PortfolioProps {
  portfolio?: SinglePortfolio;
  getPortfolio: typeof getPortfolioData;
  portfolioId?: string;
  stocks: Array<Stock>;
  getAllStocks: typeof getStocks;
  getStockMetaData: typeof getStockMetadata;
  getStockIntraDayData: typeof getIntraday;
  getTransactions: typeof getTransactions;
  cancelOpenTransaction: typeof cancelTransaction;
  refreshStockIntraDayData: typeof refreshIntraday;
  stocksLoading: boolean;
}

type PortfolioPropsWithNavigation = PortfolioProps & NavigationScreenProps;

export class PortfolioScreen extends React.Component<
  PortfolioPropsWithNavigation
> {
  static navigationOptions = { title: t('PortfoliosPage.Title') };
  constructor(props: PortfolioPropsWithNavigation) {
    super(props);
  }

  componentDidMount() {
    if (
      this.props.portfolio &&
      this.props.portfolioId &&
      this.props.portfolio.portfolioInfo.portfolio === undefined
    ) {
      this.props.getPortfolio(this.props.portfolioId);
    }
    if (this.props.stocks.length === 0) {
      this.props.getAllStocks();
    }
  }

  // Debounce function with 500 ms delay
  getStockData = debounce((stocks: PortfolioStock[]) => {
    stocks.forEach((portfolioStock) => {
      const findStock = this.props.stocks.find((stock) => {
        return stock.symbol === portfolioStock.uid;
      });

      if (findStock) {
        if (!findStock.stockInfo || !findStock.stockInfo.intraday) {
          this.props.getStockMetaData(findStock, findStock.symbol);
        }
        if (!findStock.stockInfo || !findStock.stockInfo.intraday) {
          this.props.getStockIntraDayData(findStock, findStock.symbol);
        }
      }
    });
  }, 500);

  componentDidUpdate() {
    if (this.props.portfolio && this.props.portfolio.portfolioInfo.portfolio) {
      this.getStockData(this.props.portfolio.portfolioInfo.portfolio.stocks);
    }
  }

  // Fetch portfolio and stock data when refreshed.
  refreshPortfolioAndStock = () => {
    if (this.props.portfolioId) {
      this.props.getPortfolio(this.props.portfolioId);
      this.props.getTransactions(this.props.portfolioId);
    }
    if (this.props.portfolio && this.props.portfolio.portfolioInfo.portfolio) {
      this.props.portfolio.portfolioInfo.portfolio.stocks.forEach(
        async (portfolioStock) => {
          const findStock = this.props.stocks.find((stock) => {
            return stock.symbol === portfolioStock.uid;
          });
          if (findStock) {
            this.props.refreshStockIntraDayData(findStock, findStock.symbol);
          }
        }
      );
    }
  };

  render() {
    const { portfolio, stocks, stocksLoading } = this.props;
    if (portfolio) {
      return (
        <ScrollView
          refreshControl={
            <RefreshControl
              refreshing={portfolio.portfolioInfo.refreshing}
              onRefresh={this.refreshPortfolioAndStock}
              colors={[Colors.baseColor]}
            />
          }
        >
          <Card containerStyle={cardStyles.container}>
            <PortfolioInfo
              portfolio={portfolio.portfolioInfo.portfolio}
              loading={portfolio.portfolioInfo.loading}
              error={portfolio.portfolioInfo.error}
              stocks={stocks}
            />
          </Card>
          <Card containerStyle={cardStyles.container} />
          <Card containerStyle={cardStyles.container}>
            <Holdings
              portfolio={portfolio.portfolioInfo.portfolio}
              portfolioLoading={portfolio.portfolioInfo.loading}
              error={portfolio.portfolioInfo.error}
              stocks={stocks}
              stocksLoading={stocksLoading}
            />
          </Card>
          <Card containerStyle={cardStyles.container}>
            <OpenTransactions
              getTransactions={this.props.getTransactions}
              cancelOpenTransaction={this.props.cancelOpenTransaction}
              portfolioId={this.props.portfolioId}
              transactions={portfolio.transactions.transactionListing.filter(
                (transaction) => transaction.status === 'MARKET'
              )}
              loading={portfolio.transactions.loading}
              error={portfolio.transactions.error}
            />
          </Card>
          <Card containerStyle={cardButtonStyles.cardButton}>
            <CardButton
              iconName={IconNames.events}
              translationTitle={'PortfolioPage.Events'}
              // TODO: navigate to events page
              onPress={() => this.props.navigation.navigate(RouteName.Home)}
            />
          </Card>
          <Card containerStyle={cardButtonStyles.cardButton}>
            <CardButton
              iconName={IconNames.manage}
              translationTitle={'PortfolioPage.Manage'}
              // TODO: navigae to manage portfolio page
              onPress={() => this.props.navigation.navigate(RouteName.Home)}
            />
          </Card>
        </ScrollView>
      );
    } else {
      // TODO: Format the error message to user
      return <Text>Error, portfolio not found! </Text>;
    }
  }
}

const mapStateToProps = (state: RootState) => ({
  portfolioId: state.portfolioListing.portfolioId,
  portfolio: state.portfolioListing.portfolioListing.find((portfolio) => {
    return portfolio.uid === state.portfolioListing.portfolioId;
  }),
  stocks: state.stocksListing.stocks,
  stocksLoading: state.stocksListing.loading,
});

const mapDispatchToProps = (dispatch: Dispatch) =>
  bindActionCreators(
    {
      getPortfolio: getPortfolioData,
      getAllStocks: getStocks,
      getStockMetaData: getStockMetadata,
      getStockIntraDayData: getIntraday,
      refreshStockIntraDayData: refreshIntraday,
      getTransactions: getTransactions,
      cancelOpenTransaction: cancelTransaction,
    },
    dispatch
  );
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(PortfolioScreen);
