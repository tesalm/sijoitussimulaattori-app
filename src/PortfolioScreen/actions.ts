import { Dispatch } from 'redux';

import { PortfolioApiRequest } from '../utils/api';
import { Portfolio } from './reducers';

export enum ActionType {
  RequestPortfolioBegin = '[Portfolio] Data Request',
  RequestPortfolioSuccess = '[Portfolio] Data Success',
  RequestPortfolioFailure = '[Portfolio] Data Failure',
}

export type PortfolioAction =
  | RequestPortfolioBegin
  | RequestPortfolioSuccess
  | RequestPortfolioFailure;

export class RequestPortfolioBegin {
  readonly type = ActionType.RequestPortfolioBegin;
  constructor() {
    return { type: this.type };
  }
}

export class RequestPortfolioSuccess {
  readonly type = ActionType.RequestPortfolioSuccess;
  constructor(public portfolio: Portfolio) {
    return { type: this.type, portfolio };
  }
}

export class RequestPortfolioFailure {
  readonly type = ActionType.RequestPortfolioFailure;
  constructor(public error: Error) {
    return { type: this.type, error };
  }
}

const getPortfolioData = (name: string) => async (
  dispatch: Dispatch<PortfolioAction>
) => {
  dispatch(new RequestPortfolioBegin());
  try {
    const data = await PortfolioApiRequest(name);
    dispatch(new RequestPortfolioSuccess(data));
  } catch (error) {
    dispatch(new RequestPortfolioFailure(error));
  }
};

export { getPortfolioData };
