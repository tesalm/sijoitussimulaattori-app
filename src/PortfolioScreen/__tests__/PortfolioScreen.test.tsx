import { shallow } from 'enzyme';
import * as React from 'react';
import configureStore from 'redux-mock-store';

import { getIntraday, getStockMetadata, getStocks, refreshIntraday } from '../../MarketScreen/actions';
import { getPortfolioData } from '../actions';
import { PortfolioProps, PortfolioScreen } from '../PortfolioScreen';

describe('portfolio tests', () => {
  // Mock for navigation.
  const navigationMock: any = {};
  const middlewares: any = [];
  const mockStore = configureStore(middlewares);

  const defaultStockProps: PortfolioProps = {
    portfolio: undefined,
    getPortfolio: getPortfolioData,
    name: undefined,
    loading: false,
    error: undefined,
    refreshing: false,
    stocks: [],
    getAllStocks: getStocks,
    getMeta: getStockMetadata,
    getHistoryData: jest.fn(),
    getIntra: getIntraday,
    refreshIntra: refreshIntraday,
  };

  const successStockProps: PortfolioProps = {
    portfolio: {
      name: 'Portfolio1',
      ownerId: 'klsdjfs',
      balance: 500,
      revenue: -2.4,
      totalMarketValue: 5000,
      stocks: [
        {
          symbol: 'AAPL',
          name: 'Apple',
          amount: 340,
          avgPrice: 132,
        },
      ],
    },
    stocks: [
      {
        symbol: 'AAPL',
        name: 'Apple',
        high: 375.53,
        low: 354.38,
        revenue: 3894.34,
        close: 293.42,
        currency: 'USD',
        stockInfo: {
          stockMetadata: {
            symbol: 'AAPL',
            name: 'Apple Inc.',
            type: 'string',
            region: 'string',
            marketOpen: '2018-12-12 11:20:00',
            marketClose: '2018-12-12 11:20:00',
            timeZone: 'US/Eastern',
            currency: 'USD',
            fetchTime: new Date('2018-12-17T01:24:00.000Z'),
          },
          intraday: {
            intradayQuote: [
              {
                symbol: 'AAPL',
                date: '2018-12-12T00:20:00.000Z',
                open: 170.4,
                high: 170.43,
                low: 165.43,
                close: 168.63,
                volume: 12279994,
              },
              {
                symbol: 'AAPL',
                date: '2018-12-12T00:10:00.000Z',
                open: 170.4,
                high: 170.43,
                low: 165.43,
                close: 168.63,
                volume: 12279994,
              },
            ],
            fetchTime: new Date('2018-12-17T01:24:00.000Z'),
          },
          historyData: {
            historyDataQuote: [
              {
                symbol: 'AAPL',
                date: '2018-12-15T00:00:00.000Z',
                open: 170.4,
                high: 170.43,
                low: 165.43,
                close: 162.63,
                volume: 12279994,
              },
              {
                symbol: 'AAPL',
                date: '2018-12-14T00:00:00.000Z',
                open: 170.4,
                high: 170.43,
                low: 161.43,
                close: 168.63,
                volume: 12279994,
              },
            ],
            fetchTime: new Date('2018-12-17T01:24:00.000Z'),
          },
          metaLoading: false,
          intraLoading: false,
          historyLoading: false,
          refreshing: false,
          metaError: undefined,
          intraError: undefined,
          historyError: undefined,
        },
      },
    ],
    getMeta: jest.fn(),
    getIntra: jest.fn(),
    refreshIntra: jest.fn(),
    getHistoryData: jest.fn(),
    refreshing: false,
    name: 'Portfolio1',
    getAllStocks: jest.fn(),
    getPortfolio: jest.fn(),
    loading: false,
  };

  it('Portfolioscreen renders correctly', async () => {
    const wrapper = shallow(
      <PortfolioScreen {...defaultStockProps} {...navigationMock} />,
      { context: { store: mockStore() } }
    );
    expect(wrapper.dive()).toMatchSnapshot();
  });

  it('PortfolioScreen renders correctly with portfolio and stocks', async () => {
    const wrapper = shallow(
      <PortfolioScreen {...successStockProps} {...navigationMock} />,
      { context: { store: mockStore() } }
    );
    expect(wrapper.dive()).toMatchSnapshot();
  });
});
