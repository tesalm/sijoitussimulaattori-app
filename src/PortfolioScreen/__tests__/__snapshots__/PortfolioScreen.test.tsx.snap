// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`portfolio tests PortfolioScreen renders correctly with portfolio and stocks 1`] = `
<_class
  refreshControl={
    <RefreshControlMock
      colors={
        Array [
          "#004D40",
        ]
      }
      onRefresh={[Function]}
      refreshing={false}
    />
  }
>
  <RefreshControlMock
    colors={
      Array [
        "#004D40",
      ]
    }
    onRefresh={[Function]}
    refreshing={false}
  />
  <View>
    <Themed.Card
      containerStyle={
        Object {
          "height": 130,
          "margin": 0,
        }
      }
    >
      <Component
        loading={false}
        portfolio={
          Object {
            "balance": 500,
            "name": "Portfolio1",
            "ownerId": "klsdjfs",
            "revenue": -2.4,
            "stocks": Array [
              Object {
                "amount": 340,
                "avgPrice": 132,
                "name": "Apple",
                "symbol": "AAPL",
              },
            ],
            "totalMarketValue": 5000,
          }
        }
        stocks={
          Array [
            Object {
              "close": 293.42,
              "currency": "USD",
              "high": 375.53,
              "low": 354.38,
              "name": "Apple",
              "revenue": 3894.34,
              "stockInfo": Object {
                "historyData": Object {
                  "fetchTime": 2018-12-17T01:24:00.000Z,
                  "historyDataQuote": Array [
                    Object {
                      "close": 162.63,
                      "date": "2018-12-15T00:00:00.000Z",
                      "high": 170.43,
                      "low": 165.43,
                      "open": 170.4,
                      "symbol": "AAPL",
                      "volume": 12279994,
                    },
                    Object {
                      "close": 168.63,
                      "date": "2018-12-14T00:00:00.000Z",
                      "high": 170.43,
                      "low": 161.43,
                      "open": 170.4,
                      "symbol": "AAPL",
                      "volume": 12279994,
                    },
                  ],
                },
                "historyError": undefined,
                "historyLoading": false,
                "intraError": undefined,
                "intraLoading": false,
                "intraday": Object {
                  "fetchTime": 2018-12-17T01:24:00.000Z,
                  "intradayQuote": Array [
                    Object {
                      "close": 168.63,
                      "date": "2018-12-12T00:20:00.000Z",
                      "high": 170.43,
                      "low": 165.43,
                      "open": 170.4,
                      "symbol": "AAPL",
                      "volume": 12279994,
                    },
                    Object {
                      "close": 168.63,
                      "date": "2018-12-12T00:10:00.000Z",
                      "high": 170.43,
                      "low": 165.43,
                      "open": 170.4,
                      "symbol": "AAPL",
                      "volume": 12279994,
                    },
                  ],
                },
                "metaError": undefined,
                "metaLoading": false,
                "refreshing": false,
                "stockMetadata": Object {
                  "currency": "USD",
                  "fetchTime": 2018-12-17T01:24:00.000Z,
                  "marketClose": "2018-12-12 11:20:00",
                  "marketOpen": "2018-12-12 11:20:00",
                  "name": "Apple Inc.",
                  "region": "string",
                  "symbol": "AAPL",
                  "timeZone": "US/Eastern",
                  "type": "string",
                },
              },
              "symbol": "AAPL",
            },
          ]
        }
      />
    </Themed.Card>
    <Themed.Card
      containerStyle={
        Object {
          "height": 200,
          "margin": 0,
        }
      }
    />
    <Themed.Card
      containerStyle={
        Object {
          "flex": 1,
          "margin": 0,
        }
      }
    >
      <Holdings
        loading={false}
        portfolio={
          Object {
            "balance": 500,
            "name": "Portfolio1",
            "ownerId": "klsdjfs",
            "revenue": -2.4,
            "stocks": Array [
              Object {
                "amount": 340,
                "avgPrice": 132,
                "name": "Apple",
                "symbol": "AAPL",
              },
            ],
            "totalMarketValue": 5000,
          }
        }
        stocks={
          Array [
            Object {
              "close": 293.42,
              "currency": "USD",
              "high": 375.53,
              "low": 354.38,
              "name": "Apple",
              "revenue": 3894.34,
              "stockInfo": Object {
                "historyData": Object {
                  "fetchTime": 2018-12-17T01:24:00.000Z,
                  "historyDataQuote": Array [
                    Object {
                      "close": 162.63,
                      "date": "2018-12-15T00:00:00.000Z",
                      "high": 170.43,
                      "low": 165.43,
                      "open": 170.4,
                      "symbol": "AAPL",
                      "volume": 12279994,
                    },
                    Object {
                      "close": 168.63,
                      "date": "2018-12-14T00:00:00.000Z",
                      "high": 170.43,
                      "low": 161.43,
                      "open": 170.4,
                      "symbol": "AAPL",
                      "volume": 12279994,
                    },
                  ],
                },
                "historyError": undefined,
                "historyLoading": false,
                "intraError": undefined,
                "intraLoading": false,
                "intraday": Object {
                  "fetchTime": 2018-12-17T01:24:00.000Z,
                  "intradayQuote": Array [
                    Object {
                      "close": 168.63,
                      "date": "2018-12-12T00:20:00.000Z",
                      "high": 170.43,
                      "low": 165.43,
                      "open": 170.4,
                      "symbol": "AAPL",
                      "volume": 12279994,
                    },
                    Object {
                      "close": 168.63,
                      "date": "2018-12-12T00:10:00.000Z",
                      "high": 170.43,
                      "low": 165.43,
                      "open": 170.4,
                      "symbol": "AAPL",
                      "volume": 12279994,
                    },
                  ],
                },
                "metaError": undefined,
                "metaLoading": false,
                "refreshing": false,
                "stockMetadata": Object {
                  "currency": "USD",
                  "fetchTime": 2018-12-17T01:24:00.000Z,
                  "marketClose": "2018-12-12 11:20:00",
                  "marketOpen": "2018-12-12 11:20:00",
                  "name": "Apple Inc.",
                  "region": "string",
                  "symbol": "AAPL",
                  "timeZone": "US/Eastern",
                  "type": "string",
                },
              },
              "symbol": "AAPL",
            },
          ]
        }
      />
    </Themed.Card>
    <Themed.Card
      containerStyle={
        Object {
          "height": 50,
          "margin": 0,
        }
      }
    >
      <Component />
    </Themed.Card>
    <Themed.Card
      containerStyle={
        Object {
          "height": 50,
          "margin": 0,
        }
      }
    >
      <Component />
    </Themed.Card>
  </View>
</_class>
`;

exports[`portfolio tests Portfolioscreen renders correctly 1`] = `
<_class
  refreshControl={
    <RefreshControlMock
      colors={
        Array [
          "#004D40",
        ]
      }
      onRefresh={[Function]}
      refreshing={false}
    />
  }
>
  <RefreshControlMock
    colors={
      Array [
        "#004D40",
      ]
    }
    onRefresh={[Function]}
    refreshing={false}
  />
  <View>
    <Themed.Card
      containerStyle={
        Object {
          "height": 130,
          "margin": 0,
        }
      }
    >
      <Component
        loading={false}
        stocks={Array []}
      />
    </Themed.Card>
    <Themed.Card
      containerStyle={
        Object {
          "height": 200,
          "margin": 0,
        }
      }
    />
    <Themed.Card
      containerStyle={
        Object {
          "flex": 1,
          "margin": 0,
        }
      }
    >
      <Holdings
        loading={false}
        stocks={Array []}
      />
    </Themed.Card>
    <Themed.Card
      containerStyle={
        Object {
          "height": 50,
          "margin": 0,
        }
      }
    >
      <Component />
    </Themed.Card>
    <Themed.Card
      containerStyle={
        Object {
          "height": 50,
          "margin": 0,
        }
      }
    >
      <Component />
    </Themed.Card>
  </View>
</_class>
`;
